presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java

```diff
@@ -132,34 +136,65 @@ public boolean isPlanRewritten()
         }
     }
 
-    public class CteConsumerTransformer
+    public static class CteConsumerTransformer
             extends SimplePlanRewriter<CteTransformerContext>
     {
         private final PlanNodeIdAllocator idAllocator;
 
         private final VariableAllocator variableAllocator;
 
-        public CteConsumerTransformer(PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
+        private final Session session;
+
+        private static final List<Class<? extends PlanNode>> PRECOMPUTE_PLAN_NODES = ImmutableList.of(JoinNode.class, SemiJoinNode.class, AggregationNode.class);
+
+        public CteConsumerTransformer(Session session, PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
         {
             this.idAllocator = requireNonNull(idAllocator, "idAllocator must not be null");
             this.variableAllocator = requireNonNull(variableAllocator, "variableAllocator must not be null");
+            this.session = requireNonNull(session, "Session is null");
         }
 
         @Override
         public PlanNode visitCteReference(CteReferenceNode node, RewriteContext<CteTransformerContext> context)
         {
-            context.get().addDependency(node.getCteName());
-            context.get().pushActiveCte(node.getCteName());
-            // So that dependent CTEs are processed properly
+            HashMap<String, CTEInformation> cteInformationMap = session.getCteInformationCollector().getCteInformationMap();
+            CTEInformation cteInfo = cteInformationMap.get(node.getCteId());
+            switch (getCteMaterializationStrategy(session)) {
+                case HEURISTIC_STRICT:
```

Comment:
do we want to also disable materialization of VALUES nodes? so only proceed if the query is a proper join/aggregation on top of table scans